Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Src/Training.c

```diff
@@ -1,49 +1,61 @@
 #include "Training.h"
 
-void trainingTask(uint8_t* data)
+void trainingTask (uint8_t* data)
 {
 	validData = 0;
+	int motorOneState = motorCheck (*data);
+	int motorTwoState = motorCheck (*(data + 1));
+	int lightState = lightsCheck (*(data + 2));
 	
-	if (motorCheck(*data) && motorCheck(*data) == motorCheck(*(data + 1)) && lightsCheck(*(data + 2)) ) // set outputArray if motors data are valid, motors are in sync and lights data is valid
+	if (motorOneState && motorOneState == motorTwoState) // set outputArray values of motor if motors data are valid, motors are in sync
 	{
-		for (int i = 0; i < sizeof outputArray / sizeof *outputArray; i++)
-		{
-			outputArray[i] = *(data + i);
-		}	
+		outputArray[0] = *data;
+		outputArray[1] = *(data + 1);
 	}
 	else
 	{
-		for (int i = 0; i < sizeof outputArray / sizeof *outputArray; i++)
-		{
-			outputArray[i] = 0;
-		}	
+		outputArray[0] = 0;
+		outputArray[1] = 0;
 	}
 	
-	if (motorCheck (*data))  // set first motor bit in validData to 1 if motor's data is valid
+	if (lightState) // set outputArray values of lights if lights data are valid
+	{
+		outputArray[2] = *(data + 2);
+	}
+	else
+	{
+		outputArray[2] = 0;
+	}	
+	
+	if (motorOneState)  // set first motor bit in validData to 1 if motor's data is valid
 	{
 		validData = validData | (1 << FIRSTMOTOR);
 	}
 	
-	if (motorCheck (*(data + 1))) // set second motor bit in validData to 1 if motor's data is valid
+	if (motorTwoState) // set second motor bit in validData to 1 if motor's data is valid
 	{
 		validData = validData | (1 << SECONDMOTOR);
 	}
-	if (lightsCheck (*(data+2))) // set lights bit in validData to 1 if lights data is valid
+	
+	if (lightState) // set lights bit in validData to 1 if lights data is valid
 	{
 		validData = validData | (1 << LIGHTSBOARD);
 	}	
 }
 
 int motorCheck (uint8_t testMotor)
 {
-	uint8_t motorMask = 0b11000001;
+	uint8_t forwardMask = 0b10000001;
+	uint8_t reverseMask = 0b11000000;
 	
-	if ((testMotor & motorMask) == 0b10000000) // check the case that motor is in reverse (power on, negative velocity and reverse)
+	if ((testMotor & forwardMask) == forwardMask) // check the case that motor is in reverse (power on, negative velocity and reverse)
 		return 1;
-	else if ((testMotor & motorMask) == motorMask) // check the case that motor is in forward (power on, positive velocity and forward)
+	else if ((testMotor & reverseMask) == reverseMask) // check the case that motor is in forward (power on, positive velocity and forward)
```

Comment:
Logically, this doesn't work for all cases. For the first if statement, if the velocity sign is 1, when you do the AND with the mask, the resulting bit 7 will go to zero and it will appear that everything is fine. For the second if statement, if the forward/reverse is a 1, when you do the AND with the mask, the resulting bit 0 will go to zero and it will appear that everything is fine.

Reply:
A little hint: When you create a mask, you turn on the bits you are interested in. In your case, for knowing the velocity sign, the power state, and the forward/reverse, you are interested in bits 0,6, and 7. After ANDing with the mask, you compare the result to the desired values post AND. So you would compare to your current `forwardMask` for the forward case and `reverseMask` for the reverse case